torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2247,88 +2385,220 @@ def _load_sharded_state_dict(
         with self.set_state_dict_type(StateDictType.SHARDED_STATE_DICT):
             return self.load_state_dict(state_dict, strict)
 
-    def forward(self, *args: Any, **kwargs: Any) -> Any:
+    def _reshard(
+        self,
+        params: List[FlatParameter],
+        free_full_params: bool,
+        free_mp_shard: bool,
+    ) -> None:
+        """
+        Reshards unsharded flattened parameters.
+
+        Args:
+            handles (List[FlatParamHandle]): Handles giving the flattened
+                parameters to reshard.
+            free_full_params (bool): Whether to free the unsharded flattened
+                parameter.
+            free_mp_shard (bool): Whether to free the reduced-precision sharded
+                flattened parameter.
+        """
+        if not params:
+            return
+        if free_full_params:
+            self._free_full_params(params)
+        if free_mp_shard:
+            self._free_mp_shard(params)
+        self._use_param_local_shard(params)
+
+    def _pre_forward_reshard(self, handles: List[FlatParamHandle]):
+        """
+        TODO (awgu) (linjianma): This is a naive resharding policy and should
+        probably be replaced with something more intelligent that factors in
+        the execution order parameter sequence.
+
+        Args:
+            handles (List[FlatParamHandle]): :class:`FlatParamHandle` s that
+                manage :class:`FlatParameter` s that have at least some part of
+                the module's parameters.
+        """
+        params_to_reshard: List[FlatParameter] = []
+        for handle in self._handles:
```

Comment:
and here :) it should be `handles`

Reply:
I think this is correct (given my intentions). The method argument `handles` gives which handles should be unsharded for **this** module's `forward()`. Here, we are iterating over all handles to see which are currently unsharded from **previous** modules' `forward()`s and resharding them.

For this discussion: https://github.com/awgu/pytorch/pull/4/files#r912247407
I think that passing `module_flat_param_handles` is correct.

I think the only error is the one you pointed out here: https://github.com/awgu/pytorch/pull/4/files#r912381702
We want to only unshard (all-gather) the parameters for the passed-in handles, not all handles.

Reply:
I see, now I understand the logic now. Maybe add some comment about what the input to `_pre_forward_reshard` will do? Since when I read the code I just believe that input contains handles to be resharded.

Or I think maybe do some reorganization: create a function called something like `_get_pre_forward_handles_to_reshard` and move parts of `_pre_forward_reshard` into that, and call `_reshard` after `_get_pre_forward_handles_to_reshard`?